### PR TITLE
Process pending operations on network status change

### PR DESCRIPTION
**Pull request by hovsepm**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [ ] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

A few sentences describing the overall goals of the pull request.

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m [hunk 1]
```diff
@@ -195,4 +197,43 @@ + (NSString *)tableNameForPartition:(NSString *)partition {
   return nil;
 }
 
+- (NSArray<MSPendingOperation *> *)pendingOperationsWithToken:(MSTokenResult *)token {
+  NSString *tableName = [MSDBDocumentStore tableNameForPartition:token.partition];
+  NSString *selectionQuery =
+      [NSString stringWithFormat:@"SELECT * FROM \"%@\" WHERE \"%@\" \"IS NOT NULL\"", tableName, kMSPendingOperationColumnName];
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: "IS NOT NULL" shouldn't be in quotes, it looks like this right now:

`SELECT * FROM "TABLE" WHERE "COLUMN_NAME" "IS NOT NULL"`

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m [hunk 2]
```diff
@@ -195,4 +197,43 @@ + (NSString *)tableNameForPartition:(NSString *)partition {
...
+  if (result.count == 0) {
+    return pendingDocuments;
+  }
+
+  // Create object of MSPendingOperation for each row.
+  for (id row in result) {
+    NSString *documentId = row[0][self.documentIdColumnIndex];
+    NSString *partition = row[0][self.partitionColumnIndex];
+
+    // If the document is expired, return an error and delete it.
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: It's not returning an error

```suggestion
    // If the document is expired, log a message and delete it.
```

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m [hunk 3]
```diff
@@ -195,4 +197,43 @@ + (NSString *)tableNameForPartition:(NSString *)partition {
...
+    NSDate *expirationTime = [MSUtility dateFromISO8601:row[0][self.expirationTimeColumnIndex]];
+    NSDate *currentDate = [NSDate date];
+    if ([expirationTime laterDate:currentDate] == currentDate) {
+      [self deleteWithToken:token documentId:documentId];
+      MSLogInfo([MSDataStore logTag], @"Document expired and removed from local cache: partition: %@, documentId: %@", partition,
+                documentId);
+      continue;
+    }
+
+    // TODO: verify that * 1000 for the expirationTime is valid.
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: This needs to be fixed before the PR can be merged

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDocumentStore.h [hunk 4]
```diff
@@ -69,6 +72,14 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (MSDocumentWrapper *)readWithToken:(MSTokenResult *)token documentId:(NSString *)documentId documentType:(Class)documentType;
 
+/**
+ * Get all pending operations.
+ *
+ * @param token CosmosDB token.
+ * @return List of all pending operations.
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDocumentStore.h]:
```suggestion
 *
 * @return List of all pending operations.
```

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Initialize pending operation object.
+ *
+ * @param operation Pending operation.
+ * @param partition Document partition.
+ * @param documentId Document Id.
+ * @param document Document object as string.
+ * @param etag Document etag.
+ * @param expirationTime Document expiration time.
+ * @return A pending operation instance.
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h]:
```suggestion
 *
 * @return A pending operation instance.
```

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.m [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    _document = document;
+    _etag = etag;
+    _expirationTime = expirationTime;
+  }
+  return self;
+}
+
+- (long)getDeviceTimeToLiveFromOperation {
+  long deviceTimeToLive = kMSDataStoreTimeToLiveNoCache;
+  if (self.expirationTime == -1) {
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.m]: Is there a constant for this value?  We should create one if not.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 7]
```diff
@@ -658,4 +664,117 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
   }
 }
 
+#pragma mark - Reachability
+
+- (void)networkStateChanged:(NSNotificationCenter *)__unused notification {
+  //(void)notification;
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Remove commented line.  Also what will the notification be used for in the future?

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 8]
```diff
@@ -658,4 +664,117 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
...
+    [self onNetworkGoesOnline];
+  }
+}
+
+- (void)onNetworkGoesOffline {
+}
+
+- (void)onNetworkGoesOnline {
+  @synchronized(self) {
+
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Remove extra new line

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 9]
```diff
@@ -658,4 +664,117 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
...
+                                                  }
+                                                }];
+                                       } else if ([operation.operation isEqualToString:kMSPendingOperationDelete]) {
+
+                                         // Perform delete operation
+                                         [MSDataStore
+                                             deleteWithPartition:(NSString *)operation.partition
+                                                      documentId:(NSString *)operation.documentId
+                                                    writeOptions:(MSWriteOptions * _Nullable) writeOptions
+                                               completionHandler:(MSDocumentWrapperCompletionHandler) ^ (MSDocumentWrapper *
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]:
```suggestion
                                               completionHandler:(MSDocumentWrapperCompletionHandler) ^(MSDocumentWrapper *
```

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 10]
```diff
@@ -658,4 +664,117 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
...
+                                                                                                         documentWrapper) {
+                                                 if (!documentWrapper.error) {
+                                                   [self.dataOperationProxy.documentStore deleteWithToken:tokenResponses.tokens[0]
+                                                                                               documentId:operation.documentId];
+                                                 } else if (documentWrapper.error.errorCode == MSACDocumentNotFoundErrorCode ||
```
Comment (jwallra) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: This block is the same as the one on line 733.  Perhaps consider refactoring it into a different method?

**Comment by jaelim-ms** [comment]
CI build continues failing. Seems the PR isn't quite ready for review.
1. **Are tests passing locally?** box isn't checked and there are so many unit test failures. Please make sure the PR is healthy and ready for merge after review.
2. There is no unit tests for new implementation. Please fill the gap in unit tests.

**Comment by t-nafle** [comment]
@jaelim-ms When I run unit tests locally (as a batch) most of the MSDataStoreTests fail, but then when I run each of them individually they all pass. I wonder if something is misconfigured somewhere (like the data module is disabled and never reenabled...?) or if the tests are just really flakey. 
Also yes we do need to add unit tests for the code for this PR.

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m (hunk 1 above)
Comment (jwallra) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: Please leave a comment if you resolve without making a change

Comment (jwallra) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: Ok, blame github... I went to the files page and it only showed changes from 15 commits, not all of them 🤦‍♂️

Comment (t-nafle) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: Sorry I did make a change I just didn't push, the changes are now pushed 👍

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+// Licensed under the MIT License.
+
+#import <Foundation/Foundation.h>
+
+@interface MSPendingOperation : NSObject
+
+/**
+ * Pending operation.
+ */
+@property(nonatomic, strong) NSString *operation;
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h]: Please use `copy` attribute for all `NSString` properties.

```suggestion
@property(nonatomic, copy) NSString *operation;
```

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+ */
+@property(nonatomic, strong) NSString *document;
+
+/**
+ * Document etag.
+ */
+@property(nonatomic, strong) NSString *etag;
+
+/**
+ * Document epiration time.
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h]:
```suggestion
 * Document expiration time.
```

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.m [hunk 13]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import "MSPendingOperation.h"
+#import "MSDataStore.h"
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.m]: nit: sort imports.

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.m [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  return self;
+}
+
+- (long)getDeviceTimeToLiveFromOperation {
+  long deviceTimeToLive = kMSDataStoreTimeToLiveNoCache;
+  if (self.expirationTime == kMSDataStoreTimeToLiveInfinite) {
+    deviceTimeToLive = kMSDataStoreTimeToLiveInfinite;
+  } else {
+    NSTimeInterval now = NSDate.timeIntervalSinceReferenceDate + NSTimeIntervalSince1970;
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.m]: For readability, please use `[[NSDate date] timeIntervalSince1970]` instead of computing the current time.

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.m [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    NSTimeInterval newExpirationTime = self.expirationTime - now;
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.m]: Consider rename `newExpirationTime`, because it's a diff, not a timestamp.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 16]
```diff
@@ -668,4 +673,115 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
...
+  } else {
+    MSLogInfo([MSAppCenter logTag], @"Internet connection is up.");
+    [self onNetworkGoesOnline];
+  }
+}
+
+- (void)onNetworkGoesOffline {
+}
+
+- (void)onNetworkGoesOnline {
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: May I suggest refactoring this method to make it shorter and more readable? One possible "quick" refactoring is to extract code block "Perform create and update operation" and "// Perform delete operation" into separate private methods.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 17]
```diff
@@ -668,4 +673,115 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
...
+                                                    } else {
+                                                      [self.dataOperationProxy.documentStore deleteWithToken:tokenResponses.tokens[0]
+                                                                                                  documentId:operation.documentId];
+                                                    }
+
+                                                  } else if (documentWrapper.error.errorCode == MSACDocumentNotFoundErrorCode ||
+                                                             documentWrapper.error.errorCode == MSACDocumentConflictErrorCode) {
+                                                    MSLogError(
+                                                        [MSDataStore logTag],
+                                                        @"Failed to call Cosmos delete API. Remote operation failed with error code: %ld",
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Should be "Cosmos create or update API", not "Cosmos delete API".

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 18]
```diff
@@ -668,4 +673,115 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
...
+                                                       documentId:(NSString *)operation.documentId
+                                                         document:(id<MSSerializableDocument>)operation.document
+                                                     writeOptions:(MSWriteOptions * _Nullable) writeOptions
+                                                completionHandler:^(MSDocumentWrapper *documentWrapper) {
+                                                  if (!documentWrapper.error) {
+                                                    if (writeOptions.deviceTimeToLive > kMSDataStoreTimeToLiveNoCache ||
+                                                        writeOptions.deviceTimeToLive == kMSDataStoreTimeToLiveInfinite) {
+                                                      [self.dataOperationProxy.documentStore upsertWithToken:tokenResponses.tokens[0]
+                                                                                             documentWrapper:documentWrapper
+                                                                                                   operation:nil
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Just want to make sure we don't forget to add a test covering setting `pending_operation` column to `nil` when cosmos Db was processed successfully.

Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Suggestion: It would be helpful to have a comment explain why we're setting `pending_operation` column to `nil` here.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 19]
```diff
@@ -668,4 +673,115 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
...
+
+                                         // Perform delete operation
+                                         [MSDataStore
+                                             deleteWithPartition:(NSString *)operation.partition
+                                                      documentId:(NSString *)operation.documentId
+                                                    writeOptions:(MSWriteOptions * _Nullable) writeOptions
+                                               completionHandler:(MSDocumentWrapperCompletionHandler) ^(MSDocumentWrapper *
+                                                                                                         documentWrapper) {
+                                                 if (!documentWrapper.error) {
+                                                   [self.dataOperationProxy.documentStore deleteWithToken:tokenResponses.tokens[0]
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: I don't understand this line - when we perform a DELETE and get back a cosmos Db Error, why must we delete the pending operation from the local cache?

My understanding is (implemented in Android SDK at https://github.com/Microsoft/AppCenter-SDK-Android/blob/94b91be7bed62254046bb5f28b4bf7b5e5f85afe/sdk/appcenter-storage/src/main/java/com/microsoft/appcenter/storage/Storage.java#L939-L943):
```
when we get back a cosmos Db Error
if (error code is 404 || error code is 409 || operation already expired) 
	delete the operation from local cache
else
	do nothing
```

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 20]
```diff
@@ -668,4 +673,115 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
...
+                                                    [self.dataOperationProxy.documentStore deleteWithToken:tokenResponses.tokens[0]
+                                                                                                documentId:operation.documentId];
+                                                  } else {
+                                                    MSLogError([MSDataStore logTag], @"Failed to call Cosmos create or replace API: %@",
+                                                               [documentWrapper.error.error localizedDescription]);
+                                                  }
+                                                }];
+                                       } else if ([operation.operation isEqualToString:kMSPendingOperationDelete]) {
+
+                                         // Perform delete operation
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: My understanding is we need to remove the pending operation from local cache the operation has expired. I don't see any logic in codepath the DELETE operation. Did I miss anything?

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 18 above)
Comment (dhei) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: There're quite a few edge cases in this method, we need unit tests covering those codepaths.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 20 above)
Comment (hovsepm) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: @dhei  we have if (not expatriated check) else (delete from cache via deleteWithToken method) in both places in the method that you review.

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h (hunk 11 above)
Comment (MehrdadMzfr) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h]: @dhei can you please explain?  we don't need a deep copy here

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 16 above)
Comment (MehrdadMzfr) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: addressed

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h (hunk 11 above)
Comment (dhei) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h]: Yeah it's an Objective-C convention for NSString properties, you almost always want to use `copy` attribute because of immutability. We follow this convention everywhere in this repo. See more about this convention at https://stackoverflow.com/a/10715145/873234.

File: AppCenterDataStorage/AppCenterDataStorage/Internal/MSDataStorePrivate.h [hunk 21]
```diff
@@ -56,6 +56,21 @@ static NSString *const kMSDefaultApiUrl = @"https://tokens.appcenter.ms/v0.1";
...
+/**
+ * Synchronize the document from CosmosDB with local cache.
+ *
+ * @param token The CosmosDB auth token.
+ * @param documentId The identifier of a document.
+ * @param documentWrapper Document wrapper object.
+ * @param pendingOperation Pending operation.
+ * @param operationExpirationTime oOperation expiration time.
+ */
+- (void)synchronizeLoclaCacheWithCosmosDbWithToken:(MSTokenResult *)token
```
Comment (yochaude) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/MSDataStorePrivate.h]: Typo in the name: `synchronizeLocalCacheWithCosmosDbWithToken`.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 22]
```diff
@@ -74,6 +75,8 @@ @implementation MSDataStore
 - (instancetype)init {
   if ((self = [super init])) {
     _tokenExchangeUrl = (NSURL *)[NSURL URLWithString:kMSDefaultApiUrl];
+
+    // FIXME: move that to document store (task #60212)
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: You can remove this FIXME as task https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/60212 is closed.

Comment (yochaude) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: That's not going to be very easy. This would require every method in the document store to become asynchronous. That's why I kept it here.

If not done already, I suggest you wrap the entire sync call in a block to run it on the dispatch queue.

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h [hunk 23]
```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import <Foundation/Foundation.h>
+
+@interface MSPendingOperation : NSObject
```
Comment (yochaude) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h]: Why not use MSDocumentWrapper instead? I don't see extra properties in this object.

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m [hunk 24]
```diff
@@ -98,6 +91,25 @@ - (BOOL)upsertWithToken:(MSTokenResult *)token
...
+       deviceTimeToLive:(NSInteger)deviceTimeToLive {
+  /*
+   * Compute expiration time as now + device time to live (in seconds).
+   * If device time to live is set to infinite, set expiration time as null in the database.
+   * Note: If the cache/store is meant to be disabled, this method should not even be called.
+   */
+
+  // This is the same as [[NSDate date] timeIntervalSince1970] - but saves us from allocating an NSDate.
+  NSTimeInterval now = NSDate.timeIntervalSinceReferenceDate + NSTimeIntervalSince1970;
+  NSTimeInterval expirationTime = -1;
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: It would be nice to use `kMSDataStoreTimeToLiveInfinite` constant instead of a "magic number" like `-1`, especially the constant is already used in line 107.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 22 above)
Comment (MehrdadMzfr) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: it's resolved

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m (hunk 18 above)
Comment (MehrdadMzfr) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: there is a separate task for this one to add unit test, goal is to get this in develop for rebranding

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h (hunk 23 above)
Comment (MehrdadMzfr) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/Model/MSPendingOperation.h]: we can , but this is more clear and there are more than needed offered in the Wrapper

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m (hunk 24 above)
Comment (dhei) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: Addressed in #1532.
